framework/src/utils/GriddedData.C

```diff
@@ -57,6 +57,77 @@ GriddedData::GriddedData(std::string file_name)
   parse(_dim, _axes, _grid, _fcn, _step, file_name);
 }
 
+GriddedData::GriddedData(std::vector<Real> x_axis,
+                         std::vector<Real> y_axis,
+                         std::vector<Real> z_axis,
+                         std::vector<Real> t_axis,
+                         std::vector<Real> fcn)
+{
+  unsigned int num_data_pt = 1;
+  std::vector<std::vector<Real>> grid;
+  unsigned int dim = 0;
+  _axes.resize(0);
+  // check the sizes of each axis, if one is zero, do not count it in _dim.
+  if (x_axis.size() != 0)
+  {
+    dim += 1;
+    _axes.push_back(0);
+    grid.resize(dim); // add another dimension to the grid
+    grid[dim - 1].resize(0);
+    grid[dim - 1] = x_axis;
+    num_data_pt *= grid[dim - 1].size();
+  }
+  if (y_axis.size() != 0)
+  {
+    dim += 1;
+    _axes.push_back(1);
+    grid.resize(dim); // add another dimension to the grid
+    grid[dim - 1].resize(0);
+    grid[dim - 1] = y_axis;
+    num_data_pt *= grid[dim - 1].size();
+  }
+  if (z_axis.size() != 0)
+  {
+    dim += 1;
+    _axes.push_back(2);
+    grid.resize(dim); // add another dimension to the grid
+    grid[dim - 1].resize(0);
+    grid[dim - 1] = z_axis;
+    num_data_pt *= grid[dim - 1].size();
+  }
+  if (t_axis.size() != 0)
+  {
+    dim += 1;
+    _axes.push_back(3);
+    grid.resize(dim); // add another dimension to the grid
+    grid[dim - 1].resize(0);
+    grid[dim - 1] = t_axis;
+    num_data_pt *= grid[dim - 1].size();
+  }
+
+  _dim = dim;
+  if (_dim == 0)
+    mooseError("No valid AXIS lines found by GriddedData");
+  if (num_data_pt != fcn.size())
+    mooseError("According to AXIS statements in GriddedData, number of data points is ",
+               num_data_pt,
+               " but ",
+               fcn.size(),
+               " function values were given by user");
+  setGrid(grid, fcn);
+}
+
+void
+GriddedData::setGrid(std::vector<std::vector<Real>> grid, std::vector<Real> data)
+{
+  _grid = grid;
```

Comment:
fill the _grid data directly instead of copying it in, same with _fcn

Reply:
Done.